### ISSUE TITLE
Making some animations to play in the background (GNA #5720)

### ISSUE THREAD

**Issue by wesnoth-bugs** [wesnoth/wesnoth#1064]
**[Original submission](https://gna.org/bugs/index.php?5720)** by zookeeper on 2006-04-07

While a recruited unit is fading in or when a unit is still moving, it's not possible to select another unit and issue orders to it, which makes playing much less fluid in non-accelerated mode, especially if there are lots of units to move. Such animations should not block UI actions, like selecting another unit, and should therefore be moved to play "in the background".

If an action (like moving, selecting, recruiting or capturing) will trigger an event, it can be known before the actual animation is started, and so if an event will trigger, then the animation can "block" UI actions like selecting another unit and moving it (like it does now whether an event triggers or not). This kind of behaviour is needed to prevent a situation where, for example, the player moves a unit into a location which triggers a name#moveto event, while at the same time selecting another unit which triggers a nameselect event.

(Reproduced on all)
Release: trunk
Priority: 5 - Normal
Severity: 2 - Minor

**Comment by wesnoth-bugs** [comment]
**Modified on 2008-01-20**

**martinxyz wrote:**

> From looking at the code, it already is possible to select a different unit during a move (or attack). This could trigger an event. Currently players never do this because they learned that their unit will get deselected after a move. I'm about to change this.

**Comment by wesnoth-bugs** [comment]
**Modified on 2008-01-20**

**boucman wrote:**

> I don't see how this help...
> 
> 
> 
> the code is still stuck in the animation loop during the move...
> 
> 
> 
> unfortunately it will be very tricky to have mvt anims be real background anims. This is post 1.4 stuff anyhow

**Comment by wesnoth-bugs** [comment]
**Modified on 2008-01-21**

**martinxyz wrote:**

> During the animation loop mouse events are still being processed the normal way by calling events::pump(). However most actions are blocked by the command_disabler. This is not &quot;true&quot; backgrounding, of course.

**Comment by wesnoth-bugs** [comment]
**Modified on 2008-02-02**

**boucman changed status:** None -> Postponed

**boucman wrote:**

> complete background anims is planned for 1.5

**Comment by wesnoth-bugs** [comment]
**Modified on 2010-08-28**

**eleazar wrote:**

> 1.9 has come and i still find it annoying to have to wait for the recruit animation to finish before i'm allowed to recruit/recall another unit.

**Comment by wesnoth-bugs** [comment]
**Modified on 2010-08-29**

**boucman wrote:**

> it's pretty tricky from a UI pont of view...
> 
> 
> 
> the problem is that multiple animations could take place simultaneously and we need to keep WML events in the proper order, some of the happens before animations, some happen after, and they all need to look good.
> 
> 
> 
> this is esp tricky with mvt animations

**Comment by wesnoth-bugs** [comment]
**Modified on 2010-08-31**

**eleazar wrote:**

> There's a strong argument that could be made that movement animations shouldn't be allowed to happen at the same time.  Otherwise it's unclear which hexes are occupied and which are not, which unit moved the shroud, etc.
> 
> 
> 
> The FR is just for &quot;some&quot; animations to play in the background.

**Comment by wesnoth-bugs** [comment]
**Modified on 2012-01-29**

**shadowmaster changed status:** Postponed -> None

**Comment by Wedge009** [comment]
The underlying issue behind the request doesn't appear to be an issue in 1.14 - recruitment/recalling doesn't block animations such as unit idle animations and village flag waving, nor does it block selection of other units. Not that I could tell, anyway.

**Comment by jostephd** [comment]
@Wedge009 It's possible to select unit B while unit A is moving, but it's not possible to issue orders to unit B until unit A has finished moving. For example:

1. Create three Gryphon Masters
2. Select one of them
3. Click on a hex that's 10 hexes away from the selected unit
4. The unit starts moving. Before it reaches its destination, select another Gryphon Master and click on an empty hex.

In step 4, the selection succeeds but clicking on a empty hex does nothing. It could, instead, schedule a move to after the current move finishes... and, I guess, if the _original_ move triggers an event (say, `moveto`), then the scheduled moves could be canceled, same as the "Cancel pending actions" setting in the Load Game dialog.

**Comment by Wedge009** [comment]
Ah, okay. I understand the issue better with the instructions given. Confirmed in 1.14.7.

**Comment by Pentarctagon** [comment]
Given things can happen in moveto, enter_hex, and exit_hex events, being able to move multiple units at the same time seems undesirable.

**Comment by cooljeanius** [comment]
> Given things can happen in moveto, enter_hex, and exit_hex events, being able to move multiple units at the same time seems undesirable.

Well maybe if Wesnoth detects that there are possible events of such type in the scenario code, it could transform the cursor into a little "🚷" symbol (and prevent clicking) while hovering over units while another is moving, but otherwise, if there are no such events, it could allow them? I personally have wished to be able to get other moves started while current move animations are still playing myself, personally...

**Comment by gfgtdf** [comment]
> Given things can happen in moveto, enter_hex, and exit_hex events, being able to move multiple units at the same time seems undesirable.

In theory this could be implemented in a way that ordered tasks are just intererrupted when such an event occurs, similar to how currently sighted events interrupt the current movement.

The plan by the third post and which i think is also the plan of the first post, is a bit difference since it is to run the animations asynchronously from the event executions, meaning that the animations are queued and that the event execution thus happens immidiateley (since they don't have for the animation to end), such an implementation would sure have its advantages, but i imagine it to be very difficult to implement since its simply just a huge change.

**Comment by gfgtdf** [comment]
The units wouldn't really move at the same time, the issue is just about "scheduling" the move, the actual movement wouldn't start after the first move is finished if i undestood the request correctly.
